### DISCUSSION TITLE
3rd party authentication_backend

### DISCUSSION THREAD

**Discussion by cornelinux** [authelia/authelia#8566] [Ideas]
### Description

I would like to use Authelia with 2nd factors managed in [privacyidea](https://github.com/privacyidea/privacyidea) allowing an organizsation to manage their 2nd factors centrally and use these for all login scenrios.

I was not sure, if I should comment on #2784 or #3069, but decided to open a new FR.

Since the config allows to define the `authentication_backend `:

~~~~yaml
authentication_backend:
  ldap:
        <ldap sepcific config>
~~~~

I think from a config standpoint and update resistance it should be possible to "configure" other authentication backends.
So I was wondering if there is or can be a way to add custom authentication backends.

### Use Case

An organizsation uses 2nd factors at VPN via RADIUS, at Windows Login, at Linux Login via the PAM stack and othe locations. The 2nf factors are centrally managed within privacyIDEA. (or name any other MFA management system).

The organization is also using Authelia for some Webpages, that do not allow authentication via SAML or OpenID Connect making it difficult to add a 2nd factor. So such web application (aka WebAppX) is located behind Authelia. Within Authelia a 2nd factor is managed.

The problem is that admins now need to manage 2nd factors in two places. In Authelia for access to WebAppX and in privacyIDEA for authentication at the VPN, at Windows clients, at Linux clients and all other locations.

The idea is to allow the managemen of all 2nd factors in one place and connect Authelia to this authentication system.

### Details

I would like the description of how to add a new authentication_backend to authelia including

* Which API it needs to provide
* where it needs to be located in the filesystem
* how it needs to be registered and loaded

It could then be configured like:

~~~~yaml
authentication_backend:
 my_funky_backend:
    individual_optionA: ....
    individual_optionB: ...
~~~~

The possibility to add such backend would be enough - no need to implement such backend.

### Documentation

In regards to privacyIDEA such backend would call the `validate/check` endpoint.
https://privacyidea.readthedocs.io/en/latest/modules/api/validate.html#post--validate-check

It could be simply configured like all other plugins:

~~~~yaml
authenitcation_backend:
    privacyidea:
        url: https://privacyidea.example.com/instance1
        verify_ssl: true
        realm: admins        
~~~~

### Pre-Submission Checklist

- [X] I agree to follow the [Code of Conduct](http://www.authelia.com/code-of-conduct)

- [X] I have checked for related issues and checked the documentation

**Comment by glime-ai** [comment]
Please note this response was created using generative AI leveraging previously available information from the repository. This information may be helpful however it is our strong recommendation that you carefully consider the information before blindly applying it.

---

Currently, Authelia supports two primary methods for first-factor authentication: LDAP and file-based (YAML) storage, as detailed in the configuration documentation. For second-factor authentication, Authelia supports methods like TOTP, WebAuthn, and Duo.

Regarding the integration of a custom authentication backend like privacyIDEA for managing second factors, Authelia does not natively support custom authentication backends out of the box. The configuration and system design of Authelia are built around predefined authentication methods, primarily focusing on LDAP and file-based systems for the first factor and the mentioned methods for the second factor.

To integrate privacyIDEA or any other custom authentication system, significant modifications to Authelia's source code would be required. This would involve:
1. Understanding Authelia's plugin architecture and how authentication modules are loaded and managed.
2. Developing a custom module that can interface with the privacyIDEA API, handling the specific endpoints and responses.
3. Ensuring that this module can be configured through Authelia's YAML configuration files, similar to existing authentication backends.

This kind of development would require deep knowledge of Authelia's codebase and possibly contributing to the project directly if you intend for this to be a reusable feature for others. You might consider opening a feature request on Authelia's GitHub repository or checking if there are existing requests for such functionality.

For direct integration without modifying Authelia, you might consider using privacyIDEA to manage all second factors and then using a standard protocol like SAML or OpenID Connect, which Authelia supports as an identity provider, to handle authentication requests. This setup would allow you to centralize second-factor management in privacyIDEA while leveraging Authelia for web application access control.

This comment was generated by [Glime](https://glimelab.ai/?utm_source=6fsHkcyufjQLaB&utm_medium=github).

**Comment by cornelinux** [comment]
Interesting. When I read the documentation my understanding is that Authelia does not act as a relying party but as a provider.
https://www.authelia.com/integration/openid-connect/introduction/
If Authelia was able to act as a relying party then we could redirect to another SSO IdP that supports privacyIDEA.
But since this is not the case, my search continues.

**Comment by klou** [comment]
This is exactly my desired use case.  PrivacyIdea as an already-in-place MFA provider, with Authelia also leveraging AD as a backend.

> For direct integration without modifying Authelia, you might consider using privacyIDEA to manage all second factors and then using a standard protocol like SAML or OpenID Connect, which Authelia supports as an identity provider, to handle authentication requests. This setup would allow you to centralize second-factor management in privacyIDEA while leveraging Authelia for web application access control.

This implies also setting up something like SimpleSAML to act as an intermediary ...

**Comment by james-d-elliott** [comment]
I've thought a lot about this a few times, and have discussed it with some of the team. We're of the opinion that we'll not support third parties as a authentication backend (i.e. 1FA) unless they are based on a very well established specification which has adoption by more than one client and one server application without an extremely compelling reason.

With rare exceptions we also probably will not support third party 2FA options, notable exception being Duo. The value is the alternate channel with Duo.

The likely compromises that seem reasonable from our end and that we have long had intention to support have existing feature requests (one long standing as a feature request and the other long been considered) such as [OpenID Connect 1.0 Relying Party Support](https://github.com/authelia/authelia/issues/2728) and [SCIM 2.0 Support](https://github.com/authelia/authelia/issues/7668) making this a duplicate. In essence we'd support the OpenID Connect 1.0 Relying Party role and allow administrators to configure RP clients as an alternative means for users to login. We would provide users a means by which to link their existing Authelia accounts with these providers via their iss and sub claims per the specification.

To handle 2FA when required there would logically be a single means by which we should do this; we should by default require additional 2FA,, and give administrators the option to trust the AMR values provided by the provider. We are already going down this track by re-implementing the authorization level as an AMR record, and will likely allow administrators to configure custom policies that utilize AMRs in upcoming versions.

Regardless any implementation that doesn't include a [RFC8176](https://datatracker.ietf.org/doc/html/rfc8176) `amr` claim in their provider will require additional 2FA on any endpoint that requires 2FA.

While we appreciate ideas like this especially when so much thought has been put into it, I'm closing this as a duplicate and/or rejected for the reasons noted above.
